airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams/async_job.py

```diff
@@ -0,0 +1,277 @@
+
+
+class Status(str, Enum):
+    """Async job statuses"""
+
+    COMPLETED = "Job Completed"
+    FAILED = "Job Failed"
+    SKIPPED = "Job Skipped"
+    STARTED = "Job Started"
+    RUNNING = "Job Running"
+    NOT_STARTED = "Job Not Started"
+
+
+class AsyncJob(ABC):
+    """Abstract AsyncJob base class"""
+
+    @abstractmethod
+    def start(self):
+        """Start remote job"""
+
+    @abstractmethod
+    def restart(self):
+        """Restart failed job"""
+
+    @property
+    @abstractmethod
+    def restart_number(self):
+        """Number of restarts"""
+
+    @property
+    @abstractmethod
+    def completed(self) -> bool:
+        """Check job status and return True if it is completed, use failed/succeeded to check if it was successful"""
+
+    @property
+    @abstractmethod
+    def failed(self) -> bool:
+        """Tell if the job previously failed"""
+
+    @abstractmethod
+    def update_job(self, batch=None):
+        """Method to retrieve job's status, separated because of retry handler"""
+
+    @abstractmethod
+    def get_result(self) -> Any:
+        """Retrieve result of the finished job."""
+
+
+class ParentAsyncJob(AsyncJob):
+    """Group of async jobs"""
+
+    def __init__(self, api, jobs: List[AsyncJob]):
+        self._api = api
+        self._jobs = jobs
+        self._restart_number = 0
+
+    def start(self):
+        """Start each job in the group."""
+        for job in self._jobs:
+            job.start()
+
+    def restart(self):
+        """Restart failed jobs"""
+        for job in self._jobs:
+            if job.failed:
+                job.restart()
+            self._restart_number = max(self._restart_number, job.restart_number)
+
+    @property
+    def restart_number(self):
+        """Number of restarts"""
+        return self._restart_number
+
+    @property
+    def completed(self) -> bool:
+        """Check job status and return True if all jobs are completed, use failed/succeeded to check if it was successful"""
+        return all(job.completed for job in self._jobs)
+
+    @property
+    def failed(self) -> bool:
+        """Tell if any job previously failed"""
+        return any(job.failed for job in self._jobs)
+
+    def update_job(self, batch=None):
+        """Checks jobs status in advance and restart if some failed."""
+        batch = self._api.new_batch()
+        unfinished_jobs = [job for job in self._jobs if not job.completed]
+        for jobs in chunks(unfinished_jobs, 50):
+            for job in jobs:
+                job.update_job(batch=batch)
+
+            while batch:
+                # If some of the calls from batch have failed, it returns  a new
+                # FacebookAdsApiBatch object with those calls
+                batch = batch.execute()
+
+    def get_result(self) -> Any:
+        """Retrieve result of the finished job."""
+        for job in self._jobs:
+            yield from job.get_result()
+
+    def split_job(self) -> "AsyncJob":
```

Comment:
why is this implemented here? I don't see it in `AsyncJob`

Reply:
now it is there, in theory, any job can be splited